### PR TITLE
Move our costMatrix caching to memory segments

### PR DESCRIPTION
**Pull request by chobobobo**
Also includes a potential way for us to manage memory segments.  The idea is that if a memory segment is loaded, you can choose how you wish to act on it.  For cost matrices, if the segment loads then the cache gets updated, and the segment currently only gets loaded when the codebase gets loaded.  This populates the cache stored in a global object.

The reasoning behind doing this is that we currently serialize & deserialize the costmatrix each time we use it, which carries a cost, and storing it into normal memory also carries a heavy serialize/deserialize cost.  Now the cost matrices are cached as objects in the global scope.

### CODE REVIEW

**Comment by chobobobo** [comment]
I still need to add code to make the transition smooth from the existing cache and clean it up.

File: ocsMemory.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+};
+mod.processSegments = () => {
+    const costMatrixCache = RawMemory.segments[SEGMENTS.COSTMATRIX_CACHE];
+    if (costMatrixCache) {
+    	if (DEBUG) logSystem('RawMemory', 'loading pathfinder cache..');
+        Room.loadCostMatrixCache(JSON.parse(costMatrixCache));
+        mod.deactivateSegment(SEGMENTS.COSTMATRIX_CACHE);
+    }
+};
+mod.numSaved = 0;
```
Comment (Spedwards) [inline ocsMemory.js]: Keep fields at the top of the file.

File: global.js [hunk 2]
```diff
@@ -178,6 +178,9 @@ mod.LAB_REACTIONS = {
     [RESOURCE_CATALYZED_GHODIUM_ACID]: [RESOURCE_GHODIUM_ACID, RESOURCE_CATALYST],
     [RESOURCE_CATALYZED_GHODIUM_ALKALIDE]: [RESOURCE_GHODIUM_ALKALIDE, RESOURCE_CATALYST],
 };
+mod.SEGMENTS = {
```
Comment (Spedwards) [inline global.js]: Maybe name it `MEMORY_SEGMENTS` or `MEM_SEGMENTS`. Or at the very least add a comment stating that this is for memory segments and not something else.

**Comment by chobobobo** [comment]
There's a potential problem I hadn't considered when implementing this.  We move between cloud instances quite rapidly, rarely the same instance twice, so if I wait a tick to load the cache I am likely loading it into a global on the wrong instance.

**Comment by chobobobo** [comment]
I also need to confirm that when we reload the code the current costMatrix in the segment is the most up to date.

**Comment by chobobobo** [comment]
@Spedwards I believe this meets the changes requested now, when I merged this into traveler I did notice an increase in incomplete paths, that still needs to be debugged before this is ready.

**Comment by chobobobo** [comment]
I have changes for this that make it work, but it won't merge into dev cleanly until traveler is in so it's on hold until traveler( #751 ) is merged.

**Comment by chobobobo** [comment]
Also, I need to store the deserialized cost matrices somewhere else, otherwise the benefit of caching them deserialized gets lost when I have to serialize them all again before they are stored.

**Comment by chobobobo** [comment]
This still has conflicts with #918 once that is merged it can be added to canary.

**Comment by chobobobo** [comment]
Ok, hit the limit for segments faster than I had expected, this should not be put into canary until code to resolve memory limits is implemented.  We also shouldn't be caching empty cost Matrices.

**Comment by chobobobo** [comment]
This needs to be tested when the cache moves into 2 segments before it can be merged into dev

File: global.js [hunk 3]
```diff
@@ -153,6 +153,12 @@ for(let a in REACTIONS){
         mod.LAB_REACTIONS[REACTIONS[a][b]] = [a,b];
     }
 }
+mod.MEM_SEGMENTS = {
+    COSTMATRIX_CACHE: {
+        start: 0,
+        end: 5
```
Comment (karlthepagan) [inline global.js]: would it be possible to start at 99 and work down?

**Comment by chobobobo** [comment]
Yes, I was just thinking about that myself, I will make that change in the morning

**Comment by chobobobo** [comment]
@karlthepagan please have a look, I am quite confident now that this is ready for dev, running it currently and tested transition over 100kb boundary.

**Pull request merged**